The TextFilter function cannot search for some characters

### Describe the bug
- The TextFilter function cannot search for some characters.
  - Search may not be possible in languages other than English. The impact is likely to be especially large in Asia in general.

### Steps to Reproduce
Steps to reproduce the behavior:
1. Access at https://avodemo.com/avo/resources/users
2. Press the Filter button
3. Enter the word `(` or `音楽` in the User names filter input form (音楽 means music in Japanese).
4. Click the Filter by user names button
5. javascript error and cannot search
```
stimulus.js:2017 Error invoking action "keypress->text-filter#tryToSubmit"
URIError: URI malformed
    at decodeURIComponent (<anonymous>)
    at n.b64DecodeUnicode (filter_controller.js:38:12)
    at n.changeFilter (filter_controller.js:50:33)
    at n.tryToSubmit (text_filter_controller.js:12:19)
    at tse.invokeWithEvent (stimulus.js:362:25)
    at tse.handleEvent (stimulus.js:329:18)
    at Xae.handleEvent (stimulus.js:31:25)

{identifier: 'text-filter', controller: n, element: div, index: 0, event: KeyboardEvent}controller: n {context: fse}element: divevent: KeyboardEvent {isTrusted: true, immediatePropagationStopped: false, params: {…}, key: 'Enter', stopImmediatePropagation: ƒ, …}identifier: "text-filter"index: 0[[Prototype]]: Object
handleError @ stimulus.js:2017
handleError @ stimulus.js:1423
invokeWithEvent @ stimulus.js:368
handleEvent @ stimulus.js:329
handleEvent @ stimulus.js:31
```

### Expected behavior & Actual behavior
To be able to search by text entered in text-filte.

### System configuration
**Avo version**:
3.0.6

**Rails version**:
7.1.2

**Ruby version**:
3.2.2

**License type**:
 - [ ] Community
 - [ ] Pro
 - [x] Advanced

**Are you using Avo monkey patches, overriding views or view components?**
<!-- (Mark [x] inside the brackets) -->
 - [ ] Yes. If so, please post code samples.
 - [x] No

### Screenshots or screen recordings
https://github.com/avo-hq/avo/assets/14293538/068e30b9-cb6a-49bd-9689-4df28f0211e4

https://github.com/avo-hq/avo/assets/14293538/92dc9def-88d4-4d2f-97be-af12e0d42189

#### Impact
We plan to temporarily disable the text filter for all resources to work around this issue.

 - [ ] High impact (It makes my app un-usable.)
 - [x] Medium impact (I'm annoyed, but I'll live.)
 - [ ] Low impact (It's really a tiny thing that I could live with.)

#### Urgency
 - [x] High urgency (I can't continue development without it.)
 - [ ] Medium urgency (I found a workaround, but I'd love to have it fixed.)
 - [ ] Low urgency (It can wait. I just wanted you to know about it.)

As shown below, an error occurred because filter_params contained a space, so when I replaced the space with + and added processing, I was able to search once.
However, when I tried searching again, a javascript error occurred.

```
module Avo
  module Filters
    class BaseFilter
      ...
      class << self
        def decode_filters(filter_params)
          pp "=======filter_params::#{filter_params}"
          # filter_params may contain spaces, replace with +.
          filter_params = filter_params.gsub(/[\u3000 ]/, '+')
          pp "=======filter_params::#{filter_params}"
          JSON.parse(Base64.decode64(filter_params))
        rescue
          {}
        end
```

> "=======filter_params::eyJBdm86OkZpbHRlcnM6OlRpdGxlRmlsdGVyIjoi6Z z5qW9In0="
> "=======filter_params::eyJBdm86OkZpbHRlcnM6OlRpdGxlRmlsdGVyIjoi6Z+z5qW9In0="

```
{identifier: 'text-filter', controller: n, element: div, index: 0, event: PointerEvent}
handleError @ stimulus.js:2017
handleError @ stimulus.js:1423
invokeWithEvent @ stimulus.js:368
handleEvent @ stimulus.js:329
handleEvent @ stimulus.js:31
stimulus.js:2017 Error invoking action "click->text-filter#changeFilter"

DOMException: Failed to execute 'atob' on 'Window': The string to be decoded is not correctly encoded.
    at n.b64DecodeUnicode (http://localhost:3000/avo-assets/avo.base.js:538:77532)
    at n.changeFilter (http://localhost:3000/avo-assets/avo.base.js:538:77752)
    at tse.invokeWithEvent (http://localhost:3000/avo-assets/avo.base.js:534:205536)
    at tse.handleEvent (http://localhost:3000/avo-assets/avo.base.js:534:204927)
    at Xae.handleEvent (http://localhost:3000/avo-assets/avo.base.js:534:201021)

{identifier: 'text-filter', controller: n, element: div, index: 0, event: PointerEvent}
handleError @ stimulus.js:2017
handleError @ stimulus.js:1423
invokeWithEvent @ stimulus.js:368
handleEvent @ stimulus.js:329
handleEvent @ stimulus.js:
```


https://github.com/avo-hq/avo/assets/14293538/d9c2f770-5f0f-4332-be74-a5b8a744d00b

--- a/app/controllers/avo/base_controller.rb
+++ b/app/controllers/avo/base_controller.rb
@@ -51,7 +51,6 @@ module Avo
       end
 
       # Apply filters to the current query
-
       filters_to_be_applied.each do |filter_class, filter_value|
         @query = filter_class.safe_constantize.new(
           arguments: @resource.get_filter_arguments(filter_class)

--- a/app/controllers/avo/base_controller.rb
+++ b/app/controllers/avo/base_controller.rb
@@ -51,6 +51,7 @@ module Avo
       end
 
       # Apply filters to the current query
+
       filters_to_be_applied.each do |filter_class, filter_value|
         @query = filter_class.safe_constantize.new(
           arguments: @resource.get_filter_arguments(filter_class)
@@ -357,12 +358,12 @@ module Avo
     def set_applied_filters
       reset_filters if params[:reset_filter]
 
-      @applied_filters = Avo::Filters::BaseFilter.decode_filters(fetch_filters)
+      return @applied_filters = {} if (fetched_filters = fetch_filters).blank?
+
+      @applied_filters = Avo::Filters::BaseFilter.decode_filters(fetched_filters)
 
       # Some filters react to others and will have to be merged into this
       @applied_filters = @applied_filters.merge reactive_filters
-    rescue
-      @applied_filters = {}
     end
 
     def reactive_filters

--- a/app/javascript/js/controllers/filter_controller.js
+++ b/app/javascript/js/controllers/filter_controller.js
@@ -25,17 +25,28 @@ export default class extends Controller {
     return param
   }
 
-  b64EncodeUnicode(str) {
-    // first we use encodeURIComponent to get percent-encoded UTF-8,
-    // then we convert the percent encodings into raw bytes which
-    // can be fed into btoa.
-    return encodeURIComponent(btoa(encodeURIComponent(str).replace(/%([0-9A-F]{2})/g,
-      (match, p1) => String.fromCharCode(`0x${p1}`))))
+  decode(filters) {
+    return JSON.parse(
+      new TextDecoder().decode(
+        Uint8Array.from(
+          atob(
+            decodeURIComponent(filters)
+          ), (m) => m.codePointAt(0)
+        )
+      )
+    );
   }
 
-  b64DecodeUnicode(str) {
-    // Going backwards: from bytestream, to percent-encoding, to original string.
-    return decodeURIComponent(atob(decodeURIComponent(str)).split('').map((c) => `%${(`00${c.charCodeAt(0). toString(16)}`).slice(-2)}`).join(''))
+  encode(filtered) {
+    return encodeURIComponent(
+      btoa(
+        String.fromCodePoint(
+          ...new TextEncoder().encode(
+            JSON.stringify(filtered)
+          )
+        )
+      )
+    );
   }
 
   changeFilter() {
@@ -47,7 +58,7 @@ export default class extends Controller {
 
     // Decode the filters
     if (filters) {
-      filters = JSON.parse(this.b64DecodeUnicode(filters))
+      filters = this.decode(filters)
     } else {
       filters = {}
     }
@@ -68,7 +79,7 @@ export default class extends Controller {
 
     // Encode the filters and their values
     if (filtered && Object.keys(filtered).length > 0) {
-      encodedFilters = this.b64EncodeUnicode(JSON.stringify(filtered))
+      encodedFilters = this.encode(filtered)
     }
 
     this.navigateToURLWithFilters(encodedFilters)

--- a/app/javascript/js/controllers/filter_controller.js
+++ b/app/javascript/js/controllers/filter_controller.js
@@ -29,13 +29,13 @@ export default class extends Controller {
     // first we use encodeURIComponent to get percent-encoded UTF-8,
     // then we convert the percent encodings into raw bytes which
     // can be fed into btoa.
-    return btoa(encodeURIComponent(str).replace(/%([0-9A-F]{2})/g,
-      (match, p1) => String.fromCharCode(`0x${p1}`)))
+    return encodeURIComponent(btoa(encodeURIComponent(str).replace(/%([0-9A-F]{2})/g,
+      (match, p1) => String.fromCharCode(`0x${p1}`))))
   }
 
   b64DecodeUnicode(str) {
     // Going backwards: from bytestream, to percent-encoding, to original string.
-    return decodeURIComponent(atob(str).split('').map((c) => `%${(`00${c.charCodeAt(0).toString(16)}`).slice(-2)}`).join(''))
+    return decodeURIComponent(atob(decodeURIComponent(str)).split('').map((c) => `%${(`00${c.charCodeAt(0). toString(16)}`).slice(-2)}`).join(''))
   }
 
   changeFilter() {

--- a/lib/avo/filters/base_filter.rb
+++ b/lib/avo/filters/base_filter.rb
@@ -23,8 +23,6 @@ module Avo
       class << self
         def decode_filters(filter_params)
           JSON.parse(Base64.decode64(filter_params))
-        rescue
-          {}
         end
 
         def encode_filters(filter_params)

--- a/lib/avo/filters/base_filter.rb
+++ b/lib/avo/filters/base_filter.rb
@@ -65,7 +65,9 @@ module Avo
 
       # Fetch the applied filters from the params
       def applied_filters
-        self.class.decode_filters params[PARAM_KEY]
+        return {} if (filters_from_params = params[PARAM_KEY]).blank?
+
+        self.class.decode_filters filters_from_params
       end
 
       def visible_in_view(resource: nil, parent_resource: nil)

--- a/spec/system/avo/filters/filters_spec.rb
+++ b/spec/system/avo/filters/filters_spec.rb
@@ -270,10 +270,12 @@ RSpec.describe "Filters", type: :system do
 
     let!(:team_without_members) { create :team, name: "Without Members" }
     let!(:team_with_members) { create :team, name: "With Members" }
+    let!(:team_音楽) { create :team, name: "音楽 ✓" }
 
     before do
       team_with_members.team_members << user
       team_without_members.team_members << user
+      team_音楽.team_members << user
     end
 
     let(:url) { "/admin/resources/teams?view_type=table" }
@@ -288,7 +290,7 @@ RSpec.describe "Filters", type: :system do
 
       it "filters by name" do
         visit url
-        expect(page).to have_text("Displaying 2 item")
+        expect(page).to have_text("Displaying 3 item")
 
         open_filters_menu
         fill_in "avo_filters_name_filter", with: "With Members"
@@ -302,7 +304,26 @@ RSpec.describe "Filters", type: :system do
 
         click_on "Reset filters"
         wait_for_loaded
-        expect(page).to have_text("Displaying 2 item")
+        expect(page).to have_text("Displaying 3 item")
+      end
+
+      it "filters by 音楽 ✓" do
+        visit url
+        expect(page).to have_text("Displaying 3 item")
+
+        open_filters_menu
+        fill_in "avo_filters_name_filter", with: "音楽 ✓"
+        click_on "Filter by name"
+        wait_for_loaded
+        expect(page).to have_text("Displaying 1 item")
+
+        open_filters_menu
+        expect(page).to have_text "音楽 ✓"
+        expect(page).to have_link("Reset filters")
+
+        click_on "Reset filters"
+        wait_for_loaded
+        expect(page).to have_text("Displaying 3 item")
       end
     end
   end

--- a/spec/system/avo/filters/filters_spec.rb
+++ b/spec/system/avo/filters/filters_spec.rb
@@ -270,12 +270,12 @@ RSpec.describe "Filters", type: :system do
 
     let!(:team_without_members) { create :team, name: "Without Members" }
     let!(:team_with_members) { create :team, name: "With Members" }
-    let!(:team_音楽) { create :team, name: "音楽 ✓" }
+    let!(:team) { create :team, name: "音楽 ✓" }
 
     before do
       team_with_members.team_members << user
       team_without_members.team_members << user
-      team_音楽.team_members << user
+      team.team_members << user
     end
 
     let(:url) { "/admin/resources/teams?view_type=table" }

